FeedStoreChallenge/CoreDataFeedStore.swift

```diff
@@ -7,36 +7,75 @@ import CoreData
 public final class CoreDataFeedStore: FeedStore {
 	private static let modelName = "FeedStore"
 	private static let model = NSManagedObjectModel(name: modelName, in: Bundle(for: CoreDataFeedStore.self))
-
+	
 	private let container: NSPersistentContainer
 	private let context: NSManagedObjectContext
-
+	
 	struct ModelNotFound: Error {
 		let modelName: String
 	}
-
+	
 	public init(storeURL: URL) throws {
 		guard let model = CoreDataFeedStore.model else {
 			throw ModelNotFound(modelName: CoreDataFeedStore.modelName)
 		}
-
+		
 		container = try NSPersistentContainer.load(
 			name: CoreDataFeedStore.modelName,
 			model: model,
 			url: storeURL
 		)
 		context = container.newBackgroundContext()
 	}
-
+	
 	public func retrieve(completion: @escaping RetrievalCompletion) {
-		fatalError("Must be implemented")
+		perform { context in
+			do {
+				if let coreDataFeed = try CoreDataFeed.fetch(in: context) {
+					completion(.found(feed: coreDataFeed.localFeed, timestamp: coreDataFeed.timestamp))
+				} else {
+					completion(.empty)
+				}
+			} catch {
+				completion(.failure(error))
+			}
+		}
 	}
-
+	
 	public func insert(_ feed: [LocalFeedImage], timestamp: Date, completion: @escaping InsertionCompletion) {
-		fatalError("Must be implemented")
+		perform { context in
+			do {
+				let newCache = try CoreDataFeed.newInstance(in: context)
+				
+				let value = CoreDataFeed.managedFeedSet(from: feed, in: context)
+				
+				newCache.coreDataFeedImages = value
+				newCache.timestamp = timestamp
+				
+				try context.save()
+				completion(nil)
+			} catch {
+				context.rollback()
+				completion(error)
+			}
+		}
 	}
-
+	
 	public func deleteCachedFeed(completion: @escaping DeletionCompletion) {
-		fatalError("Must be implemented")
+		perform { context in
+			do {
+				try CoreDataFeed.fetch(in: context).map(context.delete)
+				try context.save()
+				completion(nil)
+			} catch {
+				context.rollback()
+				completion(error)
+			}
+		}
+	}
+	
+	private func perform(_ action: @escaping (NSManagedObjectContext) -> Void) {
+		let context = self.context
+		context.perform { action(context) }
```

Comment:
During the course this way of having a local reference to the context to avoid capturing self inside of the `context.perform` block is taught.
The same can be achieved using capture list.
It saves one line and one variable.

No need to make this amendment, this is just for your interest.

```suggestion
		context.perform { [context] in action(context) }
```

Reply:
Done